### ISSUE TITLE
FSDP + DTensor is not working with `SHARD_GRAD_OP` + use_orig_params

### ISSUE THREAD

**Issue by bigning** [pytorch/pytorch#123272]
### 🐛 Describe the bug

Hi team, the FSDP + DTensor is not working with `SHARD_GRAD_OP`. It works with `FULL_SHARD` .  here is the error stack trace:

```
Traceback (most recent call last):
  File "/mnt/workdisk/ning/projects/fsdp_tp_test/debug_shard_grad_op_issue.py", line 110, in <module>
    mp.spawn(
  File "/usr/lib/python3/dist-packages/torch/multiprocessing/spawn.py", line 281, in spawn
    return start_processes(fn, args, nprocs, join, daemon, start_method="spawn")
  File "/usr/lib/python3/dist-packages/torch/multiprocessing/spawn.py", line 237, in start_processes
    while not context.join():
  File "/usr/lib/python3/dist-packages/torch/multiprocessing/spawn.py", line 188, in join
    raise ProcessRaisedException(msg, error_index, failed_process.pid)
torch.multiprocessing.spawn.ProcessRaisedException: 

-- Process 0 terminated with the following error:
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/torch/multiprocessing/spawn.py", line 75, in _wrap
    fn(i, *args)
  File "/mnt/workdisk/ning/projects/fsdp_tp_test/debug_shard_grad_op_issue.py", line 100, in run_fsdp_checkpoint_save_example
    loss.backward()
  File "/usr/lib/python3/dist-packages/torch/_tensor.py", line 522, in backward
    torch.autograd.backward(
  File "/usr/lib/python3/dist-packages/torch/autograd/__init__.py", line 267, in backward
    _engine_run_backward(
  File "/usr/lib/python3/dist-packages/torch/autograd/graph.py", line 690, in _engine_run_backward
    return Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
  File "/usr/lib/python3/dist-packages/torch/distributed/fsdp/_runtime_utils.py", line 693, in _pre_backward_hook
    _unshard(
  File "/usr/lib/python3/dist-packages/torch/distributed/fsdp/_runtime_utils.py", line 321, in _unshard
    ran_pre_unshard = handle.pre_unshard()
  File "/usr/lib/python3/dist-packages/torch/distributed/fsdp/_flat_param.py", line 1252, in pre_unshard
    ret = self._writeback_orig_params()
  File "/usr/lib/python3/dist-packages/torch/utils/_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
  File "/usr/lib/python3/dist-packages/torch/distributed/fsdp/_flat_param.py", line 2238, in _writeback_orig_params
    or not _same_storage(param, flat_param_tensor)
  File "/usr/lib/python3/dist-packages/torch/distributed/fsdp/_flat_param.py", line 2718, in _same_storage
    return a.untyped_storage().data_ptr() == b.untyped_storage().data_ptr()
RuntimeError: Attempted to access the data pointer on an invalid python storage.
```

You can reproduce it with 4 GPUs using the following code:
```
import os
import torch
import torch.distributed as dist
import torch.distributed.checkpoint as DCP
import torch.multiprocessing as mp
import torch.nn as nn
from typing import Callable
from torch.distributed.fsdp import FullyShardedDataParallel as FSDP
from torch.distributed.fsdp import StateDictType, FullStateDictConfig
from torch.distributed._tensor.device_mesh import init_device_mesh
from torch.distributed.fsdp.wrap import CustomPolicy
from torch.distributed._tensor import DeviceMesh, Shard, distribute_tensor, distribute_module, DTensor
from torch.distributed.fsdp.api import ShardingStrategy
from torch.distributed.checkpoint.state_dict import (StateDictOptions, get_model_state_dict)

fs_dim = 8
tp_dim = 4
input_dim = 2

class TPModel(nn.Module):
    def __init__(self):
        super().__init__()
        self.ffn1 = nn.Linear(fs_dim, tp_dim, bias=False)
    def forward(self, x):
        return self.ffn1(x)
    def fsdp_wrap_fn(self, fsdp_mesh):
        return {'device_mesh': fsdp_mesh}

class FullShardModel(nn.Module):
    def __init__(self):
        super().__init__()
        self.net1 = nn.Linear(input_dim, fs_dim, bias=False)
        self.net2 = nn.Linear(fs_dim, fs_dim, bias=False)
        self.ffn = TPModel()
    def forward(self, x):
        return self.ffn(self.net2(self.net1(x)))

def setup(rank, world_size):
    # Running on one node so master_addr is just local host
    os.environ["MASTER_ADDR"] = "localhost"
    os.environ["MASTER_PORT"] = "28000"
    # All ranks simulataneously init the process group together.
    dist.init_process_group("nccl", rank=rank, world_size=world_size)
    torch.cuda.set_device(rank)

def cleanup():
    dist.destroy_process_group()
    
def run_fsdp_checkpoint_save_example(rank, world_size, mesh_shape):
    setup(rank, world_size)
    print("set up!")
    mesh_2d = init_device_mesh(
        'cuda', mesh_shape, mesh_dim_names=("dp", "tp")
    )
    tp_mesh = mesh_2d["tp"]
    ffn_fsdp_mesh = mesh_2d["dp"]
    mesh_1d = init_device_mesh('cuda', (world_size,))
    model = FullShardModel().to(rank)
    model.ffn.ffn1.weight = torch.nn.Parameter(distribute_tensor(model.ffn.ffn1.weight, tp_mesh, [Shard(0)]))

    def tensor_to_dtensor_hook(mod, args):
        inp, = args
        return DTensor.from_local(inp, tp_mesh, [Shard(0)])
        # return distribute_tensor(inp, tp_mesh, [Shard(0)])
    
    model.ffn.register_forward_pre_hook(tensor_to_dtensor_hook)
    
    def dtensor_to_tensor_hook(mod, inp, outp):
        return outp.to_local()
    model.ffn.register_forward_hook(dtensor_to_tensor_hook)
    
    def lambda_fn(module: torch.nn.Module):
        ret = False
        if hasattr(module, '_fsdp_wrap'):
            ret = bool(module._fsdp_wrap)
        elif hasattr(module, 'fsdp_wrap_fn') and isinstance(module.fsdp_wrap_fn, Callable):
            ret = module.fsdp_wrap_fn(ffn_fsdp_mesh)
        return ret
    model = FSDP(
        model,
        #sharding_strategy=ShardingStrategy.FULL_SHARD,
        sharding_strategy=ShardingStrategy.SHARD_GRAD_OP,
        device_mesh=mesh_1d,
        use_orig_params=True,
        auto_wrap_policy=CustomPolicy(lambda_fn),
    )


    if torch.distributed.get_rank() == 3:
        print(model)

    loss_fn = nn.MSELoss()
    optim = torch.optim.Adam(model.parameters(), lr=0.1)
    optim.zero_grad()


    outp = model(torch.rand(4,input_dim,device="cuda")).sum(dim=1)

    loss = loss_fn(outp, torch.ones(4, device='cuda'))
    loss.backward()
    optim.step()


    cleanup()


if __name__ == "__main__":
    mesh_shape = (2, 2)
    world_size = mesh_shape[0]* mesh_shape[1]
    mp.spawn(
        run_fsdp_checkpoint_save_example,
        args = (world_size, mesh_shape),
        nprocs=world_size,
        join=True
    )
```

### Versions

Collecting environment information...
PyTorch version: 2.3.0.dev20240110+cu121
Is debug build: False
CUDA used to build PyTorch: 12.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.6 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.2) 9.4.0
Clang version: Could not collect
CMake version: version 3.26.3
Libc version: glibc-2.31

Python version: 3.10.13 (main, Aug 25 2023, 13:20:03) [GCC 9.4.0] (64-bit runtime)
Python platform: Linux-5.15.0-91-generic-x86_64-with-glibc2.31
Is CUDA available: True
CUDA runtime version: 12.1.105
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: NVIDIA A100-SXM4-40GB
GPU 1: NVIDIA A100-SXM4-40GB
GPU 2: NVIDIA A100-SXM4-40GB
GPU 3: NVIDIA A100-SXM4-40GB

Nvidia driver version: 535.129.03
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.9.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.9.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.9.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.9.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.9.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.9.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.9.0
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang

**Comment by bigning** [comment]
this seems work : https://github.com/pytorch/pytorch/pull/123277

**Comment by bigning** [comment]
> I think this fix is fair to land. We probably want to add a unit test.

Thanks @awgu , do you want me to add a unit test, or you can help?

**Comment by awgu** [comment]
@weifengpy can help land the fix with a unit test.

**Comment by weifengpy** [comment]
Hi @bigning , ideally fix and unit test should be landed in the same PR. do you prefer spending time to add unit test to you PR? or me taking over the PR and add a unit test?

this is the unit test we should enhance by parameterizing SHARD_GRAD_OP: https://github.com/pytorch/pytorch/blob/a4035bea5cd4b75ff681fdec13e97f6259e0901d/test/distributed/fsdp/test_fsdp_tp_integration.py#L222

**Comment by bigning** [comment]
> Hi @bigning , ideally fix and unit test should be landed in the same PR. do you prefer adding unit test by yourself? or me taking over the PR and add a unit test?
> 
> this is the unit test we should enhance by parameterizing SHARD_GRAD_OP:
> 
> https://github.com/pytorch/pytorch/blob/a4035bea5cd4b75ff681fdec13e97f6259e0901d/test/distributed/fsdp/test_fsdp_tp_integration.py#L222

that make sense. Please take over the PR and add a unit test. thanks!

**Comment by weifengpy** [comment]
> > Hi @bigning , ideally fix and unit test should be landed in the same PR. do you prefer adding unit test by yourself? or me taking over the PR and add a unit test?
> > this is the unit test we should enhance by parameterizing SHARD_GRAD_OP:
> > https://github.com/pytorch/pytorch/blob/a4035bea5cd4b75ff681fdec13e97f6259e0901d/test/distributed/fsdp/test_fsdp_tp_integration.py#L222
> 
> that make sense. Please take over the PR and add a unit test. thanks!

sounds good. will let you know once it's ready

**Comment by bigning** [comment]
@weifengpy , can we include the fix in 2.3.1 ? thanks!

**Comment by weifengpy** [comment]
> @weifengpy , can we include the fix in 2.3.1 ? thanks!

will draft a PR for cherry pick today

**Comment by weifengpy** [comment]
> @weifengpy , can we include the fix in 2.3.1 ? thanks!
@bigning I just submitted cherry-pick. if it lands, it will appear in 2.3.1: https://github.com/pytorch/pytorch/pull/126464

**Comment by atalman** [comment]
Confirmed with 2.3.1, run repro with version print included:
```
python test.py 
2.3.1+cu121
2.3.1+cu121
2.3.1+cu121
2.3.1+cu121
2.3.1+cu121
set up!
set up!
set up!
set up!
FullyShardedDataParallel(
  (_fsdp_wrapped_module): FullShardModel(
    (net1): Linear(in_features=2, out_features=8, bias=False)
    (net2): Linear(in_features=8, out_features=8, bias=False)
    (ffn): FullyShardedDataParallel(
      (_fsdp_wrapped_module): TPModel(
        (ffn1): Linear(in_features=8, out_features=4, bias=False)
      )
    )
  )
)```
